flang/runtime/descriptor.cpp

```diff
@@ -96,12 +96,19 @@ void Descriptor::Establish(const typeInfo::DerivedType &dt, void *p, int rank,
 
 OwningPtr<Descriptor> Descriptor::Create(TypeCode t, std::size_t elementBytes,
     void *p, int rank, const SubscriptValue *extent,
-    ISO::CFI_attribute_t attribute, int derivedTypeLenParameters) {
-  std::size_t bytes{SizeInBytes(rank, true, derivedTypeLenParameters)};
+    ISO::CFI_attribute_t attribute, bool addendum,
+    const typeInfo::DerivedType *dt) {
   Terminator terminator{__FILE__, __LINE__};
+  RUNTIME_CHECK(terminator, t.IsDerived() == (dt != nullptr));
+  int derivedTypeLenParameters = dt ? dt->LenParameters() : 0;
+  std::size_t bytes{SizeInBytes(rank, addendum, derivedTypeLenParameters)};
   Descriptor *result{
       reinterpret_cast<Descriptor *>(AllocateMemoryOrCrash(terminator, bytes))};
-  result->Establish(t, elementBytes, p, rank, extent, attribute, true);
```

Comment:
Note that I removed the unconditional addendum creation.  I am not sure it is a good idea, since @jeanPerier mentioned that the compiler generated code may access addendum unconditionally in some cases, but the testing is clean for the change.  Jean, please let me know about the cases that might be broken because of this.

Reply:
Would no addendum be created for an unlimited polymorphic allocatable/pointer descriptors with this change?
If not, I think you are fine.

Reply:
Do you mean in the case like this?
```
program main
  call test()
contains
  subroutine test()
    class(*), allocatable :: c(:)
    integer :: i
    i = 7
    allocate(c(2), source=i)
    select type (x => c)
    type is (integer)
       print *, x
    end select
  end subroutine test
end program main
```

The resulting descriptor for `c` does not have the addendum.

Reply:
Surely it (`c`) needs an addendum, unless its descriptor is dynamically allocated as well.  And we're going to want to have a way to distinguish an unlimited polymorphic unallocated/unassociated entity from one that has a declared type.

Reply:
This was probably confusing to say that `c` does not have the addendum.  The descriptor is allocated by the compiler, and it is allocated with the addendum.  "Does not have the addendum" here meant that after `AllocatableApplyMold` it does not have its `f18Addendum` flag set.  Let me try a different example.

Reply:
Jean, I am not sure what use-case you thought about.

I guess it would be a runtime error if Create/Establish is called for an unlimited polymorphic, and it ends up creating a descriptor without an addendum.

From the compiler side, I think, we cover all cases well.  I tried this case with a "dynamic" descriptor:
```
module types
  type t1
     class(*), allocatable :: c
  end type t1
end module types
program main
  use types
  call test()
contains
  subroutine test()
    type(t1), allocatable :: x
    allocate(x)
    allocate(x%c, source=7)
    select type (s => x%c)
    type is (integer)
       print *, s
    end select
  end subroutine test
end program main
```

The `allocate(x)` is allocating the memory based on the element size provided by the compiler, and the compiler correctly provides size of the descriptor with an addendum.  Then, the establishing of the component descriptor properly sets the `f18addendum` flag, based on the component's unlimited polymorphic type.

Reply:
Thanks for clarifying, I think it is fine then.
Every fir.load of c descriptor may create a copy that will unconditionally copy the addendum, I was worried some descriptor allocated by the runtime without addendum would make it into inlined code as the descriptor of an unlimited polymorphic, but that is not what your patch is doing and the f18addendum flag is actually never used in inlined code (it is only set or copied).